mathesar/rpc/users.py

```diff
@@ -0,0 +1,140 @@
+"""
+Classes and functions exposed to the RPC endpoint for managing mathesar users.
+"""
+from typing import Optional, TypedDict
+from modernrpc.core import rpc_method, REQUEST_KEY
+from modernrpc.auth.basic import (
+    http_basic_auth_login_required,
+    http_basic_auth_superuser_required
+)
+
+from mathesar.rpc.exceptions.handlers import handle_rpc_exceptions
+from modernrpc.exceptions import AuthenticationFailed
+from mathesar.utils.users import (
+    get_user,
+    list_users,
+    add_user,
+    update_user_info,
+    delete_user,
+    change_password,
+    revoke_password
+)
+
+
+class UserInfo(TypedDict):
+    id: int
+    username: str
+    is_superuser: bool
+    email: str
+    full_name: str
+    short_name: str
+    display_language: str
+
+    @classmethod
+    def from_model(cls, model):
+        return cls(
+            id=model.id,
+            username=model.username,
+            is_superuser=model.is_superuser,
+            email=model.email,
+            full_name=model.full_name,
+            short_name=model.short_name,
+            display_language=model.display_language
+        )
+
+
+class UserDef(TypedDict):
+    username: str
+    password: str
+    is_superuser: bool
+    email: Optional[str]
+    full_name: Optional[str]
+    short_name: Optional[str]
+    display_language: Optional[str]
+
+
+class SettableUserInfo(TypedDict):
+    username: Optional[str]
+    is_superuser: Optional[bool]
+    email: Optional[str]
+    full_name: Optional[str]
+    short_name: Optional[str]
+    display_language: Optional[str]
+
+
+@rpc_method(name='users.add')
+@http_basic_auth_superuser_required
+@handle_rpc_exceptions
+def add(*, user_def: UserDef) -> UserInfo:
+    user = add_user(user_def)
+    return UserInfo.from_model(user)
+
+
+@rpc_method(name='users.delete')
+@http_basic_auth_superuser_required
+@handle_rpc_exceptions
+def delete(*, user_id: int) -> None:
+    delete_user(user_id)
+
+
+@rpc_method(name="users.get")
+@http_basic_auth_login_required
+@handle_rpc_exceptions
+def get(*, user_id: int) -> UserInfo:
+    user = get_user(user_id)
+    return UserInfo.from_model(user)
+
+
+@rpc_method(name='users.list')
+@http_basic_auth_login_required
+@handle_rpc_exceptions
+def list_() -> list[UserInfo]:
+    users = list_users()
+    return [UserInfo.from_model(user) for user in users]
+
+
+@rpc_method(name='users.patch')
+@http_basic_auth_login_required
+@handle_rpc_exceptions
+def patch(
+    *,
+    user_id: int,
+    user_info: SettableUserInfo,
+    **kwargs
+) -> UserInfo:
+    user = kwargs.get(REQUEST_KEY).user
+    if not (user.id == user_id or user.is_superuser):
```

Comment:
Did you try to use the `set_authentication_predicate` decorator for this? I think it would be cleaner and more reusable to try to use that, or set up a custom auth decorator for the "is self or superuser" case.

Reply:
It would also ensure that if the upstream package changes the response for failure of these decorators, that this function doesn't behave differently.

Reply:
I think it is next to impossible to use `set_authentication_predicate` for testing `is_self` because, for `is_self`, we would require `user_id` which is a parameter that we pass through the request. The way django_modern_rpc is set up, the permission checks set up through `set_authentication_predicate` happen before the RPC function's `args` and `kwargs` are populated for RPC function execution.

So, even something like this doesn't work:
```
@wraps(func)
    def wrap(*args, **kwargs):
        user_id = kwargs.get('user_id')
        wrapper = auth.set_authentication_predicate(http_basic_auth_check_user, [user_is_self_or_superuser, user_id])
        return wrapper(func)(*args, **kwargs)
```
I also tried getting the `user_id` using the request, but `request.POST` results in `None` when printed inside `http_basic_auth_check_user()`.

I finally settled on writing a custom decorator but without using `set_authentication_predicate`.